src/app/(proper_react)/(redesign)/(authenticated)/admin/churn-subscribers/ChurnAdmin.tsx

```diff
@@ -0,0 +1,121 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use client";
+
+import { useEffect, useState } from "react";
+import { useSession } from "next-auth/react";
+import styles from "./ChurnAdmin.module.scss";
+import { SubscriberChurnRow } from "knex/types/tables";
+import { getAllChurns, upsertAllChurns } from "./actions";
+
+export const ChurnAdmin = () => {
+  const session = useSession();
+  const [subscriberData, setSubscriberData] = useState<
+    SubscriberChurnRow[] | null
+  >(null);
+
+  // Load data on component mount
+  useEffect(() => {
+    const loadSubscriberData = async () => {
+      const data = await getAllChurns();
+      setSubscriberData(data);
+    };
+
+    void loadSubscriberData();
+  }, []);
+
+  const handleFileUpload = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const file = event.target.files?.[0];
+    if (!file) return;
+
+    const reader = new FileReader();
+    reader.onload = async (e) => {
+      const text = e.target?.result as string;
+      const rows = text.split("\n");
+
+      console.log({ rows: rows.length });
+      const parsedData: SubscriberChurnRow[] = rows
+        .slice(1)
```

Comment:
Maybe do a check here that the headers and number of columns match what we expect, and bail out if not? In case someone selects the wrong file.

Reply:
added